File: scripts/take_screenshots.py
#!/usr/bin/env python3
"""Playwright screenshots of the customer page: both screens, in a real browser (the Chrome that is already installed).

    scripts/run_demo.sh                                       # one terminal: the app on port 8765, real agent only
    python scripts/take_screenshots.py                        # another: writes demo/screenshots/01-...png
    python scripts/take_screenshots.py --base http://127.0.0.1:8790 --out demo/screenshots

It walks the flow a customer would: empty upload screen, the documents read, a problem with a document, the first chat message, an answer with its
suggestions, the same answer with "Show more" open, and a phone-sized view. Like the demo script it refuses the offline stand-in unless you pass
--allow-offline (for testing this script), because a screenshot of the stand-in would misrepresent ARC.
Needs:  pip install -r requirements-dev.txt
"""
import argparse
import json
import sys
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))
from tests.pdfmaker import edit  # noqa: E402

from playwright.sync_api import expect, sync_playwright  # noqa: E402

DOCS = ROOT / "demo" / "documents"
ALL = sorted(str(p) for p in DOCS.glob("*.pdf"))
WAIT_MS = 100000      # the real agent needs 15 to 25 seconds per answer; expect() has its own (short) default timeout


def main():
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--base", default="http://127.0.0.1:8765")
    ap.add_argument("--out", default=str(ROOT / "demo" / "screenshots"))
    ap.add_argument("--allow-offline", action="store_true")
    a = ap.parse_args()
    base, out = a.base.rstrip("/"), Path(a.out)
    try:
        health = json.load(urllib.request.urlopen(base + "/health", timeout=5))
    except Exception:  # noqa: BLE001
        sys.exit(f"Cannot reach {base}. Start it first with scripts/run_demo.sh")
    if not health.get("live") and not a.allow_offline:
        sys.exit("The server is the OFFLINE stand-in, not the real agent. Start it with scripts/run_demo.sh (or pass --allow-offline to test this script).")
    out.mkdir(parents=True, exist_ok=True)
    import tempfile
    tmp = Path(tempfile.mkdtemp(prefix="arc-shots-"))                                    # the bill "in another name" for the problem screen; removed at the end
    wrong_bill = tmp / "hospital_bill.pdf"
    wrong_bill.write_bytes(edit((DOCS / "hospital_bill.pdf").read_bytes(), {"Rohan Verma": "Rohan Varma"}))
    saved = []

    def see(locator, text):
        expect(locator).to_contain_text(text, timeout=WAIT_MS)

    def shot(page, name, full=True):
        page.wait_for_timeout(350)
        path = out / name
        page.screenshot(path=str(path), full_page=full)
        saved.append(path)
        print("saved", path.relative_to(ROOT) if path.is_relative_to(ROOT) else path)

    with sync_playwright() as p:
        browser = p.chromium.launch(channel="chrome", headless=True)
        ctx = browser.new_context(viewport={"width": 1100, "height": 800})
        page = ctx.new_page()
        page.set_default_timeout(100000)
        try:
            # 1. the upload screen, empty
            page.goto(base + "/")
            expect(page.locator("#checklist li")).to_have_count(10)
            shot(page, "01-upload.png")

            # 2. after "Use sample documents": ticks, the missing prescription, the way in
            page.click("#use-sample")
            see(page.locator("#ready-text"), "We recognised 9 of 10 documents")
            shot(page, "02-upload-documents-read.png")

            # 3. a document that does not fit (a bill in another name): plain reasons and a way to replace it
            page.goto(base + "/")
            page.set_input_files("#file-input", [f for f in ALL if not f.endswith("hospital_bill.pdf")] + [str(wrong_bill)])
            expect(page.locator("#attention")).to_be_visible()
            shot(page, "03-upload-needs-attention.png")

            # 4. the chat: the first message is the claim summary, with three suggestions
            page.goto(base + "/")
            page.click("#use-sample")
            expect(page.locator("#ready")).to_be_visible()
            page.click("#continue")
            expect(page.locator("#thread .chip")).to_have_count(3)
            shot(page, "04-chat-first-message.png")

            # 5. an answer: a short summary, and three new suggestions under it
            page.get_by_role("button", name="How much will be paid?").click()
            see(page.locator("#thread .msg-arc").nth(1), "₹1,22,125")
            expect(page.locator("#thread .chip")).to_have_count(3, timeout=WAIT_MS)
            shot(page, "05-chat-answer.png")

            # 6. the same answer with "Show more" open and one policy reference expanded
            answer = page.locator("#thread .msg-arc").nth(1)
            answer.get_by_role("button", name="Show more").click()
            answer.locator(".ref-chip").nth(2).click()
            shot(page, "06-chat-show-more.png")
            answer.get_by_role("button", name="Show less").click()

            # 7. a follow-up from a suggestion
            page.get_by_role("button", name="Why was my room rent reduced?").click()
            see(page.locator("#thread .msg-arc").nth(2), "62.5%")
            shot(page, "07-chat-room-rent.png")
        finally:
            ctx.close()

        # 8. a phone
        phone = browser.new_context(viewport={"width": 390, "height": 844}, device_scale_factor=2, is_mobile=True, has_touch=True)
        pp = phone.new_page()
        pp.set_default_timeout(100000)
        pp.goto(base + "/")
        pp.click("#use-sample")
        expect(pp.locator("#ready")).to_be_visible()
        pp.click("#continue")
        pp.get_by_role("button", name="How much will be paid?").click()
        see(pp.locator("#thread .msg-arc").nth(1), "₹1,22,125")
        shot(pp, "08-phone-chat.png", full=False)
        phone.close()
        browser.close()
    import shutil
    shutil.rmtree(tmp, ignore_errors=True)
    print(f"\nSaved {len(saved)} screenshots in {out}" + ("" if health.get("live") else "  (OFFLINE stand-in: test only, do not publish)"))


if __name__ == "__main__":
    main()

File: scripts/take_screenshots.mjs
#!/usr/bin/env node
// Screenshots of the 8 demo steps in docs/DEMO.md, taken from the OFFICER console (/officer) with headless Chrome. No dependencies (Node 22+ has WebSocket and fetch).
//
//   scripts/run_demo.sh                                        # in one terminal: the app on port 8765, real agent only
//   node scripts/take_screenshots.mjs                          # in another: writes docs/screenshots/step1-....png ... step8-....png
//   node scripts/take_screenshots.mjs --base http://127.0.0.1:8765 --out docs/screenshots
//
// The questions come from the table in docs/DEMO.md (terminal A = no claim, B = TC07, C = TC02), so this cannot drift from the demo script.
// It refuses to run against the offline stand-in (GET /health must say live) unless you pass --allow-offline (for testing the script itself).
// Each screenshot shows the header (with the Live agent badge), the claim panel, and the question with its answer as an officer sees it.
import { spawn } from "node:child_process";
import { mkdirSync, mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const args = process.argv.slice(2);
const opt = (name, dflt) => (args.includes(name) ? args[args.indexOf(name) + 1] : dflt);
const BASE = opt("--base", "http://127.0.0.1:8765").replace(/\/$/, "");
const OUT = path.resolve(ROOT, opt("--out", "docs/screenshots"));
const CHROME = opt("--chrome", "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome");
const PORT = Number(opt("--debug-port", "9333"));
const WIDTH = 1280;
const ANSWER_TIMEOUT_MS = 100000;
const CLAIM_FOR_TERMINAL = { A: "", B: "TC07", C: "TC02" };
const SLUGS = ["knee-replacement", "settlement-ratio", "waiting-period-dates", "assess-tc07", "why-room-rent", "missing-documents", "what-if-room-rent", "not-payable-tc02"];

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

function demoSteps() {
  const md = readFileSync(path.join(ROOT, "docs", "DEMO.md"), "utf8");
  const steps = [...md.matchAll(/^\| ([1-8]) \| ([A-C]) \| `([^`]+)` \|/gm)].map((m) => ({ n: Number(m[1]), claim: CLAIM_FOR_TERMINAL[m[2]], q: m[3] }));
  if (steps.length !== 8) throw new Error(`expected 8 steps in docs/DEMO.md, found ${steps.length}`);
  return steps;
}

class Page {
  constructor(ws) {
    this.ws = ws;
    this.id = 0;
    this.pending = new Map();
    ws.addEventListener("message", (e) => {
      const msg = JSON.parse(e.data);
      if (msg.id && this.pending.has(msg.id)) {
        const { resolve, reject } = this.pending.get(msg.id);
        this.pending.delete(msg.id);
        msg.error ? reject(new Error(msg.error.message)) : resolve(msg.result);
      }
    });
  }
  send(method, params = {}) {
    const id = ++this.id;
    return new Promise((resolve, reject) => {
      this.pending.set(id, { resolve, reject });
      this.ws.send(JSON.stringify({ id, method, params }));
    });
  }
  async eval(expression) {
    const r = await this.send("Runtime.evaluate", { expression, awaitPromise: true, returnByValue: true });
    if (r.exceptionDetails) throw new Error(r.exceptionDetails.exception?.description || r.exceptionDetails.text);
    return r.result.value;
  }
  async until(expression, what, timeoutMs = 30000) {
    const t0 = Date.now();
    while (Date.now() - t0 < timeoutMs) {
      if (await this.eval(expression)) return;
      await sleep(300);
    }
    throw new Error(`timed out waiting for ${what}`);
  }
  async shot(file, { hideOlder = true } = {}) {
    if (hideOlder) {   // only the newest question and answer, so each picture is about one step (element.style is allowed by the page's CSP; an injected <style> is not)
      await this.eval(`[...document.getElementById('transcript').children].slice(0, -2).forEach((c) => { c.dataset.shotHid = '1'; c.style.display = 'none'; })`);
    }
    await this.eval("window.scrollTo(0, 0)");
    // The page stretches to fill the viewport (footer at the bottom), so scrollHeight only ever grows. Add up the real content instead.
    const h = Math.min(3400, Math.max(640, await this.eval(`(() => { const q = (s) => document.querySelector(s); const vis = (e) => (e && !e.hidden ? e.offsetHeight : 0);
      return Math.ceil(vis(q('.site-header')) + vis(q('#offline-banner')) + 40 + Math.max(q('.side').offsetHeight, q('.chat').offsetHeight) + vis(q('.site-footer')) + 4); })()`)));
    await this.send("Emulation.setDeviceMetricsOverride", { width: WIDTH, height: h, deviceScaleFactor: 1, mobile: false });
    await sleep(250);
    const { data } = await this.send("Page.captureScreenshot", { format: "png", captureBeyondViewport: false });
    writeFileSync(file, Buffer.from(data, "base64"));
    await this.eval("document.querySelectorAll('[data-shot-hid]').forEach((c) => { c.style.display = ''; delete c.dataset.shotHid; })");
  }
}

async function launchChrome(userDir) {
  const proc = spawn(CHROME, [`--remote-debugging-port=${PORT}`, `--user-data-dir=${userDir}`, "--headless=new", "--hide-scrollbars", "--no-first-run",
    "--no-default-browser-check", "--disable-gpu", `--window-size=${WIDTH},900`, "about:blank"], { stdio: "ignore" });
  for (let i = 0; i < 60; i++) {
    try {
      const targets = await (await fetch(`http://127.0.0.1:${PORT}/json/list`)).json();
      const page = targets.find((t) => t.type === "page");
      if (page) return { proc, wsUrl: page.webSocketDebuggerUrl };
    } catch { /* not up yet */ }
    await sleep(250);
  }
  proc.kill();
  throw new Error("Chrome did not start. Is it installed at " + CHROME + " ? (use --chrome <path>)");
}

async function main() {
  const health = await fetch(`${BASE}/health`).then((r) => r.json()).catch(() => null);
  if (!health) throw new Error(`Cannot reach ${BASE}. Start it first with scripts/run_demo.sh`);
  if (!health.live && !args.includes("--allow-offline")) {
    throw new Error("The server is the OFFLINE stand-in, not the real agent. Start it with scripts/run_demo.sh (or pass --allow-offline to test this script).");
  }
  const steps = demoSteps();
  mkdirSync(OUT, { recursive: true });
  const userDir = mkdtempSync(path.join(tmpdir(), "arc-shots-"));
  const { proc, wsUrl } = await launchChrome(userDir);
  const ws = new WebSocket(wsUrl);
  await new Promise((resolve, reject) => { ws.addEventListener("open", resolve); ws.addEventListener("error", reject); });
  const page = new Page(ws);
  const written = [];
  try {
    await page.send("Page.enable");
    await page.send("Emulation.setDeviceMetricsOverride", { width: WIDTH, height: 900, deviceScaleFactor: 1, mobile: false });
    await page.send("Page.navigate", { url: BASE + "/officer" });
    await page.until("!!document.getElementById('send') && !document.getElementById('send').disabled", "the page to be ready");
    let claim = "";
    for (const step of steps) {
      if (step.claim !== claim) {   // pick the claim in the dropdown, exactly as an officer would
        await page.eval(`(() => { const s = document.getElementById('claim-select'); s.value = ${JSON.stringify(step.claim)}; s.dispatchEvent(new Event('change')); })()`);
        await page.until("!document.getElementById('send').disabled && /(is loaded|Policy questions only)/.test(document.getElementById('status').textContent)", "the claim to load");
        claim = step.claim;
      }
      const before = await page.eval("document.querySelectorAll('.answer').length");
      const clicked = await page.eval(`(() => { const b = [...document.querySelectorAll('.q-btn')].find(x => x.title === ${JSON.stringify(step.q)}); if (b) b.click(); return !!b; })()`);
      if (!clicked) throw new Error(`step ${step.n}: no quick-question button for: ${step.q}`);
      await page.until(`document.querySelectorAll('.answer').length > ${before} && !document.querySelector('.loading')`, `the answer to step ${step.n}`, ANSWER_TIMEOUT_MS);
      const status = await page.eval("(() => { const a = [...document.querySelectorAll('.answer')].pop(); return a.classList.contains('warn') || a.classList.contains('error') ? 'NOT OK' : 'ok'; })()");
      if (status !== "ok") throw new Error(`step ${step.n} did not produce a normal answer (timeout or error card); not saving a misleading screenshot`);
      await sleep(400);
      const file = path.join(OUT, `step${step.n}-${SLUGS[step.n - 1]}.png`);
      await page.shot(file);
      written.push(file);
      console.log(`step ${step.n}: ${path.relative(ROOT, file)}   (${step.q.slice(0, 60)})`);
      if (step.n === 4) {   // the same answer with every detail section open and one citation chip expanded
        await page.eval(`(() => { const a = [...document.querySelectorAll('.answer')].pop(); a.querySelector('.link-toggle')?.click(); a.querySelectorAll('.chip')[2]?.click(); })()`);
        await sleep(300);
        const f2 = path.join(OUT, "step4b-details-expanded.png");
        await page.shot(f2);
        written.push(f2);
        console.log(`step 4 (expanded): ${path.relative(ROOT, f2)}`);
        await page.eval(`(() => { const a = [...document.querySelectorAll('.answer')].pop(); a.querySelector('.link-toggle')?.click(); a.querySelectorAll('.chip')[2]?.click(); })()`);
      }
    }
  } finally {
    ws.close();
    proc.kill();
    await sleep(600);   // let Chrome finish writing to its profile folder before it is removed
    try { rmSync(userDir, { recursive: true, force: true }); } catch { /* a temp folder; the OS will clean it */ }
  }
  console.log(`\nSaved ${written.length} screenshots in ${path.relative(ROOT, OUT)}${health.live ? "" : "  (OFFLINE stand-in: test only, do not publish)"}`);
}

main().catch((e) => { console.error("\nERROR:", e.message); process.exit(1); });
